### ISSUE TITLE
[C++20][Modules] Make using STL header units practical

### ISSUE THREAD

**Issue by hugoam** [llvm/llvm-project#62241]
Hello,

It seems that currently, producing and using STL header units to be used by `mymodule` only works with stdlibc++, and the only way to use them is to:
- precompile them manually through a somewhat arcane command line (that's probably ok)
- and then specify each single one of them to the compiler through `-fmodule-file=vector.pcm`, for *each* header used by `mymodule` (that's quite unpractical for real world usage)

Is that a correct assessment of how those are supposed to be used?

If yes, it would seem that the only practical way to compile *anything* that uses STL header units is to shove one `-fmodule-file=` option in the command line for all STL headers everywhere, and pray that this won't incur too much overhead if some or most of those are not used?

Is there any plans to go in a direction in which this limitation would be somehow lifted?
Or was the vision for that problem, that it was supposed to be handled by the tooling? Would that mean the tooling is expected to first scan modules for imported header units, and then compile them accordingly?

Thanks !

**Comment by dwblaikie** [comment]
Generally it's expected that build systems would need to scan source to find the module dependencies, build/compile any referenced external (system installed .cppm files) modules and pass their built form (.pcm) to the relevant actions of the build, yes.

**Comment by dwblaikie** [comment]
(the alternative non-scan form is https://github.com/urnathan/libcody that is more on-demand approach to dependency discovery and building)

**Comment by hugoam** [comment]
> Generally it's expected that build systems would need to scan source to find the module dependencies, build/compile any referenced external (system installed .cppm files) modules and pass their built form (.pcm) to the relevant actions of the build, yes

I understand the argument, but the assumption that bothers me here is the assumption that standard library headers should be considered as regular "userland" module dependencies. The pre-module expectation is that the standard library is provided and handled by the compiler, transparently and it "just works"

When I write `#include <vector>` in a .cpp and feed that to a compiler, I don't expect to have to do anything specific to make my code work aside from just compiling the file with any compiler with a pretty basic command line

Similarly, I would expect to be able to write `import <vector>;` in a module and compile that code with a basic command line without having to or to rely on an entire order of magnitude of complexity, and a tooling infrastructure separate to the compiler

I understand this might not be just a Clang issue, but a C++20 design issue, but it's a pretty huge one in my opinion 😭 

About the overhead, is there an overhead to literally feeding all standard library modules that exist through `-fmodule-file` option, always?
Basically the way I would expect/hope it to work is `TU is asking for header unit, do we have that in the provided modules -> yes/no`, and the compiler to automatically find the corresponding module if I could specify just a *folder* where everything is nicely labeled and stored, instead of every single module separately

**Comment by ChuanqiXu9** [comment]
The header units' interfaces are still highly experimental. I mean they can change if the build systems/toolings offer a sensible requirement. 

BTW, although I don't know the contents of the talk, I think this (https://schedule.cppnow.org/session/c-modules-the-challenges-of-implementing-header-units/) may be a good summary for the issue.

> Is that a correct assessment of how those are supposed to be used?

So this is not **supposed** to be used. It is just status quo. And I am not sure how should it be supported ideally.

> Is there any plans to go in a direction in which this limitation would be somehow lifted?

I know some ideas but I don't know about a specific plan.

---

Aside of the issue, I remember there was someone who complained the status of header units to me. But he became happy after he knew about std modules. We're implementing std modules in https://reviews.llvm.org/D144994. And I think `import std;` will be much better than `import <iostream>;` or other STL headers. So this may be a better choice for you.

**Comment by hugoam** [comment]
> Aside of the issue, I remember there was someone who complained the status of header units to me too. But he became happy after he knew about std modules. We're implementing std modules in https://reviews.llvm.org/D144994. And I think `import std;` will be much better than `import <iostream>;` or other STL headers. So this may be a better choice for you.

Hmm yeah maybe I'll follow that advice for now

Initially I preferred `import <iostream>;` because it's guaranteed to have the same granularity on all compilers, while standard library modules have a compiler specific delimitation and granularity which either forces you to import *everything*, or write multiple compiler-specific code. I guess `import std;` should compile with all compilers as long as you don't try to go more fine grained

It does feel like the granularity of `import <iostream>;` could be more efficient than importing all of the STL, but I don't actually know whether that holds any truth given the implementations, maybe you can tell me more about that is true or not in Clang (or maybe I'll just benchmark it, hehe)

**Comment by ChuanqiXu9** [comment]
> It does feel like the granularity of import <iostream>; could be more efficient than importing all of the STL, but I don't actually know whether that holds any truth given the implementations, maybe you can tell me more about that is true or not in Clang (or maybe I'll just benchmark it, hehe)

For the efficiency issue, `import std;` shouldn't be slower than a `#include <iostream>` in a hello world example. This is done by lazy loading so that the size of the module may not matter so much. This is the basic requirement for modules  (although this is not specified in the doc) otherwise people wouldn't love to use it.

And benchmarking is still very welcomed since the performance of modules may not be so ideal in complex cases, e.g., in the long chained situation.

Besides the performance issue, std module is much better than `import <STL>;` since it has better semantics. Users don't need to remember concrete headers any more.
